### PR TITLE
New TorusKnotBufferGeometry

### PR DESCRIPTION
**Pull request by Mugen87**
This PR introduces `TorusKnotBufferGeometry`. Here is the corresponding [test fiddle](https://jsfiddle.net/fuwpLv10/7/).

There are some things i want to discuss:
- I'm afraid that the parameter `radialSegments` and `tubularSegments` have interchanged meanings in `TorusGeometry` and `TorusKnotGeometry`. I suggest that the implementation of `TorusGeometry` is correct, where `tubularSegments` determines the amount of vertical segments of the tube and `radialSegments` the amount of horizontal segments. Maybe it's good to align this in `TorusKnotBufferGeometry`.
- The `heightScale` parameter is confusing. It scales the z-component of a position on the torus curve which is quite extraordinary. Other geometries of three.js don't have such a parameter. And why is there no `widthScale` and `depthScale`? This implementation is somehow inconsistent. Is it basically possible to remove `heightScale` or mark it as deprecated?

Besides, i've added the `arc` parameter so `TorusKnotBufferGeometry` and  `TorusBufferGeometry` have more similar interfaces.

### CODE REVIEW

**Comment by WestLangley** [comment]
>I'm afraid that the parameter radialSegments and tubularSegments have interchanged meaning

Oops. Also look at `THREE.TubeGeometry`. And `THREE.CylinderGeometry` for that matter. There have been a lot of cooks in the kitchen. : - )

>Is it basically possible to remove heightScale or mark it as deprecated?

Yes, deprecate with a warning. That is what `BufferGeometry.scale()` is for.

>i've added the arc parameter so TorusKnotBufferGeometry and TorusBufferGeometry have more similar interfaces.

I do not think a torus knot needs an arc parameter.

**Comment by Mugen87** [comment]
The last commit removes the new `arc` parameter and marks `heightScale` as deprecated (with warning). Besides, i changed the meanings `radialSegments` and `tubularSegments` and adjusted the corresponding examples and docs.

**Comment by WestLangley** [comment]
Changing the variable names is one thing, but you are changing the constructor signatures. This will cause a lot of breakage....

**Comment by Mugen87** [comment]
I see. But that means, the interfaces of `TorusKnotBufferGeometry` and `TorusBufferGeometry` will differ. Anyway, i will undone the last commit and just rename the variables...

File: src/loaders/ObjectLoader.js [hunk 1]
```diff
@@ -225,15 +225,15 @@ THREE.ObjectLoader.prototype = {
...
 
 					case 'TorusKnotGeometry':
+					case 'TorusKnotBufferGeometry':
 
-						geometry = new THREE.TorusKnotGeometry(
+						geometry = new THREE[ data.type ](
 							data.radius,
 							data.tube,
 							data.radialSegments,
 							data.tubularSegments,
```
Comment (Mugen87) [inline src/loaders/ObjectLoader.js]: How should we handle this case? If we change `data.radialSegments` and `data.tubularSegments` things could break again, right?

**Comment by WestLangley** [comment]
Unfortunately, this is a mess. I wonder if we could use only these terms: `segments` along the path or tube, `radialSegments` around the tube circumference. Just a thought.

File: src/loaders/ObjectLoader.js (hunk 1 above)
Comment (WestLangley) [inline src/loaders/ObjectLoader.js]: Just do what you think is the least disruptive. There are no right answers here.

**Comment by Mugen87** [comment]
@WestLangley What do you think is the best way to document such conventions? Is there a place in the wiki for this (something like "coding conventions")?

**Comment by Mugen87** [comment]
I've added a vertexNormalHelper ot the [fiddle](https://jsfiddle.net/fuwpLv10/7/). Now i have to think at a hedgehog when i see the many lines on the torus :smile:

File: src/extras/geometries/TorusKnotBufferGeometry.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	this.parameters = {
+		radius: radius,
+		tube: tube,
+		tubularSegments: tubularSegments,
+		radialSegments: radialSegments,
+		p: p,
+		q: q
+	};
+
+	if( heightScale !== undefined ) console.warn( 'THREE.TorusKnotGeometry: heightScale has been deprecated. Use .scale( x, y, z ) instead.' );
```
Comment (mrdoob) [inline src/extras/geometries/TorusKnotBufferGeometry.js]: Considering that this is a new class that no one has used yet we can ignore `heightScale`. We should definitely keep the check in `TorusKnowGeometry` though.

Comment (Mugen87) [inline src/extras/geometries/TorusKnotBufferGeometry.js]: Ok! I've just implemented your feedback.

**Comment by mrdoob** [comment]
Many thanks!

**Pull request merged**